pkg/codegen/go/gen.go

```diff
@@ -1323,6 +1332,102 @@ func (pkg *pkgContext) genConfig(w io.Writer, variables []*schema.Property) erro
 	return nil
 }
 
+// genResourceModule generates a ResourceModule definition and the code to register an instance thereof with the
+// Pulumi runtime. The generated ResourceModule supports the deserialization of resource references into fully-
+// hydrated Resource instances. If this is the root module, this function also generates a ResourcePackage
+// definition and its registration to support rehydrating providers.
+func (pkg *pkgContext) genResourceModule(w io.Writer) {
+	contract.Assert(len(pkg.resources) != 0)
+
+	// Check for provider-only modules.
+	var provider *schema.Resource
+	if providerOnly := len(pkg.resources) == 1 && pkg.resources[0].IsProvider; providerOnly {
+		provider = pkg.resources[0]
+	} else {
+		imports := codegen.NewStringSet(
+			"github.com/blang/semver",
+			"github.com/pulumi/pulumi-kubernetes/sdk/v2/go/kubernetes",
+			"github.com/pulumi/pulumi/sdk/v2/go/pulumi",
+		)
+
+		pkg.genHeader(w, []string{"fmt"}, imports)
+
+		fmt.Fprintf(w, "type module struct {\n")
+		fmt.Fprintf(w, "\tversion semver.Version\n")
+		fmt.Fprintf(w, "}\n\n")
+
+		fmt.Fprintf(w, "func (m *module) Version() semver.Version {\n")
+		fmt.Fprintf(w, "\treturn m.version\n")
+		fmt.Fprintf(w, "}\n\n")
+
+		fmt.Fprintf(w, "func (m *module) Construct(ctx *pulumi.Context, name, typ, urn string) (r pulumi.Resource, err error) {\n")
+		fmt.Fprintf(w, "\tswitch typ {\n")
+
+		for _, r := range pkg.resources {
+			if r.IsProvider {
+				contract.Assert(provider == nil)
+				provider = r
+				continue
+			}
+
+			fmt.Fprintf(w, "\tcase %q:\n", r.Token)
+			fmt.Fprintf(w, "\t\tr, err = New%s(ctx, name, nil, pulumi.URN_(urn))\n", resourceName(r))
+		}
+		fmt.Fprintf(w, "\tdefault:\n")
+		fmt.Fprintf(w, "\t\treturn nil, fmt.Errorf(\"unknown resource type: %%s\", typ)\n")
+		fmt.Fprintf(w, "\t}\n\n")
+		fmt.Fprintf(w, "\treturn\n")
+		fmt.Fprintf(w, "}\n\n")
+
+		fmt.Fprintf(w, "func init() {\n")
+		fmt.Fprintf(w, "\tversion, err := %s.PkgVersion()\n", pkg.pkg.Name)
+		fmt.Fprintf(w, "\tif err != nil {\n")
+		fmt.Fprintf(w, "\t\tpanic(fmt.Errorf(\"failed to determine package version: %%v\", err))\n")
+		fmt.Fprintf(w, "\t}\n")
+		fmt.Fprintf(w, "\tpulumi.RegisterResourceModule(\n")
+		fmt.Fprintf(w, "\t\t%q,\n", pkg.pkg.Name)
+		fmt.Fprintf(w, "\t\t%q,\n", tokenToModule(pkg.resources[0].Token))
+		fmt.Fprintf(w, "\t\t&module{version},\n")
+		fmt.Fprintf(w, "\t)\n")
+		fmt.Fprintf(w, "}\n")
+	}
+
+	if provider != nil {
+		imports := codegen.NewStringSet(
+			"github.com/blang/semver",
+			"github.com/pulumi/pulumi/sdk/v2/go/pulumi",
+		)
+
+		pkg.genHeader(w, []string{"fmt"}, imports)
```

Comment:
What does the generated code look like for a provider that doesn't have any sub-modules (like `random`)? I would expect the same file to have calls to both `RegisterResourcePackage` and `RegisterResourceModule`. Like https://github.com/pulumi/pulumi-random/blob/master/sdk/python/pulumi_random/__init__.py

Reply:
Yeah, it looks like I don't have this quite right yet. https://github.com/pulumi-bot/pulumi-random/compare/f7e21f1..c596e19#diff-572719db95a5ff9f62d8b4d256f04462e146ae584e1d4612f39702aac242e4b3